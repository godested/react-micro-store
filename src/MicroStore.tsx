import React, {
  createContext,
  useReducer,
  ReactNode,
  Reducer,
  Context,
} from 'react';
import { noop, identity } from './utils';

export interface DefaultContextProvider {
  state: unknown;
  dispatch: Function;
}

export interface MicroStoreProviderProps {
  initialState?: Record<string, any>;
  reducer?: Reducer<unknown, unknown>;
  context?: Context<DefaultContextProvider>;
  children: ReactNode;
}

export const MicroStoreContext = createContext<DefaultContextProvider>({
  state: {},
  dispatch: noop,
});

export function MicroStoreProvider(props: MicroStoreProviderProps) {
  const {
    initialState = {},
    reducer = identity,
    children,
    context,
    ...restProps
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);

  const StoreContext = context || MicroStoreContext;

  return (
    <StoreContext.Provider value={{ state, dispatch }} {...restProps}>
      {children}
    </StoreContext.Provider>
  );
}
