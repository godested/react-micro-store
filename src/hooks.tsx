import { useContext, Context } from 'react';
import { MicroStoreContext, DefaultContextProvider } from './MicroStore';

export function useMicroStore<
  P extends Function,
  D extends Function,
  C extends Context<DefaultContextProvider>
>(
  mapStateToProps?: P,
  mapDispatchToProps?: D,
  context?: C
): Record<string, any> {
  const { state, dispatch } = useContext(context || MicroStoreContext);

  const mappedState = mapStateToProps ? mapStateToProps(state) : state;
  const mappedDispatch = mapDispatchToProps ? mapDispatchToProps(dispatch) : {};

  return {
    ...mappedState,
    ...mappedDispatch,
  };
}
