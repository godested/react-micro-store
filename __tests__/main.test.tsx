import React from 'react';
import renderer from 'react-test-renderer';
import { useMicroStore, MicroStoreProvider } from '../src';

function MockComponent() {
  const { foo } = useMicroStore(state => ({ foo: state.foo }));

  return <div>{foo}</div>;
}

it('should call useMicroStoreHook', () => {
  const component = renderer.create(
    <MicroStoreProvider initialState={{ foo: 'foo' }}>
      <MockComponent />
    </MicroStoreProvider>
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
