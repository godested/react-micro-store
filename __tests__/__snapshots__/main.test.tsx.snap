// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call useMicroStoreHook 1`] = `
<div>
  foo
</div>
`;
